# Market Basket Analysis on the Groceries 2.0

* Importing the libraries

In [12]:
import pandas as pd
import re
from pyECLAT import ECLAT

* Importing the Dataset

In [5]:
df=pd.read_csv('store_data (1).csv',header=None)

# Data Analysis

In [6]:
df.head()

0          1           2                 3             4   \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs               NaN           NaN   
2        chutney        NaN         NaN               NaN           NaN   
3         turkey    avocado         NaN               NaN           NaN   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1               NaN   NaN             NaN           NaN           NaN   
2               NaN   NaN             NaN           NaN           NaN   
3               NaN   NaN             NaN           NaN           NaN   
4               NaN   NaN             NaN           NaN           NaN   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1             NaN        NaN    NaN    NaN            NaN     NaN   
2             NaN        NaN    NaN    NaN            NaN     NaN   
3             NaN        NaN    NaN    NaN            NaN     NaN   
4             NaN        NaN    NaN    NaN            NaN     NaN   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                NaN              NaN      NaN        NaN  
2                NaN              NaN      NaN        NaN  
3                NaN              NaN      NaN        NaN  
4                NaN              NaN      NaN        NaN

In [4]:
df.tail()

shrimp            almonds      avocado  vegetables mix green grapes  \
7495    butter         light mayo  fresh bread             NaN          NaN   
7496   burgers  frozen vegetables         eggs    french fries    magazines   
7497   chicken                NaN          NaN             NaN          NaN   
7498  escalope          green tea          NaN             NaN          NaN   
7499      eggs    frozen smoothie  yogurt cake  low fat yogurt          NaN   

     whole weat flour yams cottage cheese energy drink tomato juice  \
7495              NaN  NaN            NaN          NaN          NaN   
7496        green tea  NaN            NaN          NaN          NaN   
7497              NaN  NaN            NaN          NaN          NaN   
7498              NaN  NaN            NaN          NaN          NaN   
7499              NaN  NaN            NaN          NaN          NaN   

     low fat yogurt green tea honey salad mineral water salmon  \
7495            NaN       NaN   NaN   NaN           NaN    NaN   
7496            NaN       NaN   NaN   NaN           NaN    NaN   
7497            NaN       NaN   NaN   NaN           NaN    NaN   
7498            NaN       NaN   NaN   NaN           NaN    NaN   
7499            NaN       NaN   NaN   NaN           NaN    NaN   

     antioxydant juice frozen smoothie spinach  olive oil  
7495               NaN             NaN     NaN        NaN  
7496               NaN             NaN     NaN        NaN  
7497               NaN             NaN     NaN        NaN  
7498               NaN             NaN     NaN        NaN  
7499               NaN             NaN     NaN        NaN

Here the NaN values refer's to 0, which means that the person didn't bought that


* Unique elements present in the df

In [9]:
unique_values = set()
for col in df:
    unique_values.update(df[col])

In [10]:
unique_values

{' asparagus',
 'almonds',
 'antioxydant juice',
 'asparagus',
 'avocado',
 'babies food',
 'bacon',
 'barbecue sauce',
 'black tea',
 'blueberries',
 'body spray',
 'bramble',
 'brownies',
 'bug spray',
 'burger sauce',
 'burgers',
 'butter',
 'cake',
 'candy bars',
 'carrots',
 'cauliflower',
 'cereals',
 'champagne',
 'chicken',
 'chili',
 'chocolate',
 'chocolate bread',
 'chutney',
 'cider',
 'clothes accessories',
 'cookies',
 'cooking oil',
 'corn',
 'cottage cheese',
 'cream',
 'dessert wine',
 'eggplant',
 'eggs',
 'energy bar',
 'energy drink',
 'escalope',
 'extra dark chocolate',
 'flax seed',
 'french fries',
 'french wine',
 'fresh bread',
 'fresh tuna',
 'fromage blanc',
 'frozen smoothie',
 'frozen vegetables',
 'gluten free bar',
 'grated cheese',
 'green beans',
 'green grapes',
 'green tea',
 'ground beef',
 'gums',
 'ham',
 'hand protein bar',
 'herb & pepper',
 'honey',
 'hot dogs',
 'ketchup',
 'light cream',
 'light mayo',
 'low fat yogurt',
 'magazines',
 'mashe

In [11]:
len(unique_values)

121

# ECLAT Implementaion

* Crating the instace which will replace the 1's with values and the rest with 0's

In [13]:
eclat_instance = ECLAT(data=df, verbose=True)

100%|██████████| 120/120 [00:00<00:00, 799.33it/s]


* Displaying the df

In [15]:
eclat_instance.df_bin.head()

avocado  salmon  mushroom cream sauce  bacon  shampoo  babies food  \
0        1       1                     0      0        0            0   
1        0       0                     0      0        0            0   
2        0       0                     0      0        0            0   
3        1       0                     0      0        0            0   
4        0       0                     0      0        0            0   

   vegetables mix  barbecue sauce  salt  fromage blanc  ...  green grapes  \
0               1               0     0              0  ...             1   
1               0               0     0              0  ...             0   
2               0               0     0              0  ...             0   
3               0               0     0              0  ...             0   
4               0               0     0              0  ...             0   

   tomatoes  burger sauce  shrimp  olive oil  mashed potato  sparkling water  \
0         0             0       1          1              0                0   
1         0             0       0          0              0                0   
2         0             0       0          0              0                0   
3         0             0       0          0              0                0   
4         0             0       0          0              0                0   

   blueberries  almonds  mineral water  
0            0        1              1  
1            0        0              0  
2            0        0              0  
3            0        0              0  
4            0        0              1  

[5 rows x 120 columns]

* Columns

In [16]:
eclat_instance.df_bin.columns

Index(['avocado', 'salmon', 'mushroom cream sauce', 'bacon', 'shampoo',
       'babies food', 'vegetables mix', 'barbecue sauce', 'salt',
       'fromage blanc',
       ...
       'green grapes', 'tomatoes', 'burger sauce', 'shrimp', 'olive oil',
       'mashed potato', 'sparkling water', 'blueberries', 'almonds',
       'mineral water'],
      dtype='object', length=120)

Here the length is 120, but the previous unique values length was 121 which includes 'NaN' also

* list with all the names of the different items

In [17]:
eclat_instance.uniq_

['avocado',
 'salmon',
 'mushroom cream sauce',
 'bacon',
 'shampoo',
 'babies food',
 'vegetables mix',
 'barbecue sauce',
 'salt',
 'fromage blanc',
 'burgers',
 'milk',
 'frozen smoothie',
 'tea',
 'spinach',
 'napkins',
 'light cream',
 'cream',
 'asparagus',
 'magazines',
 'mayonnaise',
 'muffins',
 'antioxydant juice',
 'water spray',
 'gums',
 'butter',
 'soda',
 'tomato sauce',
 'shallot',
 'tomato juice',
 'fresh tuna',
 'frozen vegetables',
 'chocolate',
 ' asparagus',
 'chicken',
 'oatmeal',
 'energy drink',
 'light mayo',
 'rice',
 'sandwich',
 'melons',
 'chili',
 'white wine',
 'toothpaste',
 'eggplant',
 'strong cheese',
 'eggs',
 'honey',
 'cake',
 'dessert wine',
 'herb & pepper',
 'cottage cheese',
 'yogurt cake',
 'chocolate bread',
 'cauliflower',
 'carrots',
 'whole wheat pasta',
 'candy bars',
 'black tea',
 'bug spray',
 nan,
 'zucchini',
 'chutney',
 'whole weat flour',
 'brownies',
 'strawberries',
 'ground beef',
 'whole wheat rice',
 'ham',
 'hot dogs',
 'min

* Length of the Unique values

In [18]:
len(eclat_instance.uniq_ )

121

* Fitting in the model

In [19]:
get_ECLAT_indexes, get_ECLAT_supports = eclat_instance.fit_all(min_support=0.2,
                                                           separator=' & ',
                                                           verbose=True)

Combination 1 by 1


1it [00:00, 34.51it/s]

Stopping in combination 1 by 1. All supports = 0.
